### DISCUSSION TITLE
"brew leaves" and "brew deps" seem to disagree

### DISCUSSION THREAD

**Discussion by candlerb** [Homebrew/discussions#2759] [Everyday usage] [answered]
I have a question about `brew leaves`.

I thought that `brew leaves` would show *all* packages which are not dependencies of another package - in other words, "top level" packages that can be safely removed.

However, I find that there are other packages which are reported by "brew deps" as not being dependencies of other packages, but not listed by "brew leaves" either.  Here's an example:

```
$ brew list | grep assimp
assimp
$ brew leaves | grep assimp
$ brew leaves -r | grep assimp
$ brew leaves -p | grep assimp
$ brew deps --installed | grep assimp
assimp:
$ 
```

So it's unclear to me whether I can safely remove `assimp` or not.  And if I can, why was it not listed by `brew leaves`?

I don't ever remember installing "assimp" so it was presumably installed as a dependency of some other package which has since been uninstalled.  I only happened to notice this because `brew outdated` tells me it wants to update it.  I'd rather uninstall it if it's not in use.

Previously I'd been tidying up by uninstalling everything listed by `brew leaves -p`, repeating until empty, which it is now:

```
$ brew leaves -p | wc -l
       0
$ 
```

FYI, `brew update` is fully up-to-date.  `brew doctor` gives nothing relevant:

```
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  six
$ 
```

Host is Macbook Pro 13" 2015, Monterey 12.1

```
$ uname -a
Darwin MacBook-Pro-4.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64 x86_64
```

Thanks!

**Comment by carlocab** [comment]
What's the output of `brew uses --installed assimp` and `brew autoremove --dry-run`?

**Comment by candlerb** [comment, marked as answer]
Ah, I just discovered this:

```
$ brew uses --installed assimp
poppler                                                             qt
$ 
```

I find that `brew leaves` does show "poppler", but not "qt".

Also:

```
$ brew deps --installed --tree
...
assimp
...

poppler
└── qt
    ├── assimp
...

qt
├── assimp
...
```

That's pretty clear.  What's confusing me then is that `brew deps --installed` (without `--tree`) doesn't show assimp as a dependency of anything, and especially not qt.

```
$ brew deps --installed | grep assimp
assimp:
$ brew deps --installed | grep ^qt:
qt: brotli ca-certificates dbus double-conversion freetype gdbm gettext giflib glib hunspell icu4c jasper jpeg libb2 libffi libpng libtiff md4c mpdecimal openssl@1.1 pcre pcre2 python@3.9 readline webp xz zstd
$ brew deps --installed qt
brotli
ca-certificates
dbus
double-conversion
freetype
gdbm
gettext
giflib
glib
hunspell
icu4c
jasper
jpeg
libb2
libffi
libpng
libtiff
md4c
mpdecimal
openssl@1.1
pcre
pcre2
python@3.9
readline
webp
xz
zstd
$ 
```

**Reply by candlerb** [reply]
Also weird:

```
$ brew deps --installed qt >a
$ brew deps qt >b
$ diff -u a b
--- a	2022-01-10 11:45:56.000000000 +0000
+++ b	2022-01-10 11:46:02.000000000 +0000
@@ -1,3 +1,4 @@
+assimp
 brotli
 ca-certificates
 dbus
@@ -22,6 +23,7 @@
 pcre2
 python@3.9
 readline
+sqlite
 webp
 xz
 zstd
```

But:

```
$ brew deps --tree --installed qt >a
$ brew deps --tree qt >b
$ diff -u a b
$
```

**Reply by candlerb** [reply]
> and `brew autoremove --dry-run`?

We crossed over :-)  No output from that one:

```
$ brew autoremove --dry-run
$ 
```

**Reply by candlerb** [reply]
Something else that *might* be related:

```
$ brew outdated
assimp (5.1.5) < 5.1.6
sqlite (3.37.1) < 3.37.2
mpv (0.34.0) != 0.34.1
$ 
```

**Reply by carlocab** [reply]
Yea, that's definitely related. I'm pretty sure these discrepancies will go away when you do `brew upgrade`.

**Reply by candlerb** [reply]
Hmm.  Does this mean: "`brew deps --installed foo` shows the dependencies of foo, but excluding those dependencies whose installed version doesn't match the latest available version" ??

If so, I wasn't expecting that.

**Reply by carlocab** [reply]
It's a bit unexpected, but it's a key part of dependency resolution (which does often have a lot of unexpected consequences).

The idea here is that you want `brew install foo` to make sure that all of its dependencies are also up-to-date, since we don't know if `foo` works with older versions of the dependencies. This means that `brew` considers having an old version of a dependency installed as not satisfying `foo`'s dependencies. The benefit of this is that `brew install foo` also upgrades and outdated but installed dependencies.

You can also see this by doing `brew info qt` (assuming you haven't yet upgraded `sqlite` and `assimp`). You'll see ✘s next to `sqlite` and `assimp` since those dependencies are currently not satisfied.

**Reply by candlerb** [reply]
> You can also see this by doing `brew info qt` (assuming you haven't yet upgraded `sqlite` and `assimp`). You'll see ✘s next to `sqlite` and `assimp` since those dependencies are currently not satisfied.

I do indeed, and thanks for your explanation:

```
$ brew info qt
qt: stable 6.2.2 (bottled), HEAD
Cross-platform application and UI framework
https://www.qt.io/
/usr/local/Cellar/qt/6.2.2 (12,156 files, 509.3MB) *
  Poured from bottle on 2022-01-08 at 13:29:30
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/qt.rb
License: GFDL-1.3-only and GPL-2.0-only and GPL-3.0-only and LGPL-2.1-only and LGPL-3.0-only
==> Dependencies
Build: cmake ✘, ninja ✘, node ✔, pkg-config ✔
Required: assimp ✘, brotli ✔, dbus ✔, double-conversion ✔, freetype ✔, glib ✔, hunspell ✔, icu4c ✔, jasper ✔, jpeg ✔, libb2 ✔, libpng ✔, libtiff ✔, md4c ✔, pcre2 ✔, python@3.9 ✔, sqlite ✘, webp ✔, zstd ✔
==> Requirements
Build: Xcode ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 44,846 (30 days), 191,813 (90 days), 712,139 (365 days)
install-on-request: 22,886 (30 days), 94,684 (90 days), 267,760 (365 days)
build-error: 76 (30 days)
```

**Comment by gromgit** [comment]
Weird. What's the output of `brew config`?

**Reply by candlerb** [reply]
> Weird. What's the output of `brew config`?

```
$ brew config
HOMEBREW_VERSION: 3.3.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 96137bc19e68398ebbb7033379df288cd8b9a3f9
Last commit: 3 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: f9fa359703cc9534e2582a4da7f32084c6b803b0
Core tap last commit: 6 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.vL4B8EjlIf/org.xquartz:0
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit broadwell
Clang: 13.0.0 build 1300
Git: 2.32.0 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.1-x86_64
CLT: 13.2.0.0.1.1638488800
Xcode: 13.2.1
$ 
```
